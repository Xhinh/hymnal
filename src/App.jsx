import React, { Component, Suspense } from "react";
import CssBaseline from "@material-ui/core/CssBaseline";
import BottomNav from "./components/BottomNav";
import Nav from "./components/TopNav";
import styles from "./App.module.scss";
import { Redirect } from "react-router-dom";

const PictureHeader = React.lazy(() => import("./components/PictureHeader"));

const Song = {
    number: Number,
    title: String,
    verse: Array(String),
    chorus: String,
    author: String
};

export const SongType = Song;

export const MainContext = React.createContext();

class App extends Component {
    state = {
        width: window.innerWidth,
        path: "/",
        pages: {
            HOME: "/",
            INDEX: "/songs",
            FAVOURITES: "/favourites",
            HISTORY: "/history",
            SETTINGS: "/settings"
        },
        title: "",
        subtitle: "",
        activeIndex: 0,
        navigate: false,
        songDisplay: "",
        songList: Array(Song),
        listLoaded: false,
        filteredList: Array(<div key={1} />),
        setProp: props => {
            this.setState({ ...props });
        },
        changePath: path => {
            this.changePath(path);
        }
    };

    changePath(path) {
        this.setState({
            navigate: !this.navigate,
            path: path
        });
    }

    componentWillMount() {
        window.addEventListener("resize", this.handleWindowSizeChange);
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this.handleWindowSizeChange);
    }

    componentDidUpdate() {
        if (this.state.navigate === true) {
            this.setState({ navigate: false });
        }
    }

    handleWindowSizeChange = () => {
        this.setState({ width: window.innerWidth });
    };

    render() {
        const { width, navigate } = this.state;
        const isMobile = width <= 960;

        if (navigate) {
            return <Redirect to={this.state.path} push={true} />;
        }

        return (
            <div className={styles.app}>
                <CssBaseline />
                <MainContext.Provider value={this.state}>
                    <section className={styles.app_body}>
                        <Nav />

                        <main className={styles.app_inner}>
                            <Suspense fallback={<div>Loading...</div>}>
                                <PictureHeader
                                    title={this.state.title}
                                    subtitle={this.state.subtitle}
                                />
                            </Suspense>
                            {this.props.children}
                        </main>

                        {/* Bottom Navigation on mobile */ isMobile ? <BottomNav /> : null}
                    </section>
                </MainContext.Provider>
            </div>
        );
    }
}

export default App;
