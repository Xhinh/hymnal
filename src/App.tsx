import React, { useReducer, useEffect } from "react";
import { Route, Switch, Redirect, useLocation } from "react-router-dom";
import { useSongLoader } from "components/CustomHooks";
import { isMobile } from "react-device-detect";
import { Spinner } from "@chakra-ui/react";
import styles from "./App.module.scss";
import { withSuspense } from "helpers";

const Header = withSuspense(React.lazy(() => import("components/Header")));
const BottomNav = withSuspense(React.lazy(() => import("components/BottomNav")));
const PictureHeader = withSuspense(React.lazy(() => import("components/PictureHeader")));

const Home = withSuspense(React.lazy(() => import("pages/Home")));
const Songs = withSuspense(React.lazy(() => import("pages/Songs")));
const Settings = withSuspense(React.lazy(() => import("pages/Settings")));

const pages = {
	HOME: "/home",
	INDEX: "/songs",
	FAVOURITES: "/favourites",
	SETTINGS: "/settings",
};

type ACTIONTYPE =
	| { type: "setTitle"; payload: string }
	| { type: "setSubtitle"; payload: string }
	| { type: "setWidth"; payload: number };

type State = {
	title: string;
	subtitle: string;
	width: number;
};

type CTX = {
	dispatch?: React.Dispatch<ACTIONTYPE>;
	songs?: Song[];
	pages?: typeof pages;
	meta?: State;
};

const initialAppState = {
	title: "",
	subtitle: "",
	width: document.body.getBoundingClientRect().width,
};

function reducer(state: State, action: ACTIONTYPE): State {
	switch (action.type) {
		case "setTitle":
			return { ...state, title: action.payload };
		case "setSubtitle":
			return { ...state, subtitle: action.payload };
		case "setWidth":
			return { ...state, width: action.payload };
		default:
			return state;
	}
}

const ScrollRestoration = () => {
	const { pathname } = useLocation();
	useEffect(() => {
		window.scrollTo(0, 0);
	}, [pathname]);
	return null;
};

export const MainContext = React.createContext<CTX>({});

function App() {
	const [state, dispatch] = useReducer(reducer, initialAppState);
	const songs = useSongLoader();

	function handleOrientationChange() {
		dispatch({ type: "setWidth", payload: document.body.getBoundingClientRect().width });
	}

	useEffect(() => {
		window.addEventListener("orientationchange", handleOrientationChange, {
			capture: false,
			passive: true,
		});
		return function cleanup() {
			window.addEventListener("orientationchange", handleOrientationChange);
		};
	}, []);

	return (
		<MainContext.Provider value={{ meta: state, dispatch, songs, pages }}>
			<div className={styles.root}>
				<section className={styles.app_body}>
					<Header />
					<ScrollRestoration />
					<main className={styles.app_inner}>
						<PictureHeader />

						<div className={styles.wrapper}>
							<Switch>
								<Route path="/home" component={Home} />
								<Route path="/songs">
									{songs.length > 1 ? (
										<Songs />
									) : (
										<div className="loader">
											<Spinner size="xl" />
										</div>
									)}
								</Route>
								<Route path="settings" component={Settings} />
								<Route>
									<Redirect to="/home" />
								</Route>
							</Switch>
						</div>
					</main>

					{/* Bottom Navigation on mobile */ isMobile ? <BottomNav /> : null}
				</section>
			</div>
		</MainContext.Provider>
	);
}

export default App;
