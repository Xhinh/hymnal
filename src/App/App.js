import React, { Component } from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import BottomNav from '../components/BottomNav/BottomNav';
import Nav from '../components/TopNav/TopNav';
import styles from './App.module.scss';
import Index from '../pages/Index/Index';
import HomePage from '../pages/Home/Home';

class App extends Component {
	constructor() {
		super();
		
		this.pages = {
			HOME: 'Home',
			INDEX: 'Index',
			FAVOURITES: 'Favourites',
			SETTINGS: 'Settings'
		};

		this.state = {
			currentPage: this.pages.HOME,
			width: window.innerWidth
		};
	}

	componentWillMount() {
		window.addEventListener('resize', this.handleWindowSizeChange);
	}

	componentWillUnmount() {
		window.removeEventListener('resize', this.handleWindowSizeChange);
	}

	handleWindowSizeChange = () => {
		this.setState({ width: window.innerWidth });
	};

	changePage(page) {
		this.setState({ currentPage: page});
	}

	getComponent() {
		let component = null;
		switch (this.state.currentPage) {
			case this.pages.HOME:
				component = <HomePage 
								navigate={this.changePage.bind(this)} 
								pages={this.pages}>
							</HomePage>;
				break;
			case this.pages.INDEX:
				component = <Index />;
				break;
			case this.pages.FAVOURITES:
				break;
			case this.pages.SETTINGS:
                break;
            default:
                break;

		}
		return component;
	}

	render() {
		const { width } = this.state;
		const isMobile = width <= 960;
		return (
			<div className={styles.app}>
				<CssBaseline />
				<section className={styles.app_body}>
					{/* Header */}
					<Nav />
					<main className={styles.app_inner}>
						{/* Main */}
						{this.getComponent()}
					</main>
					{/* Bottom Navigation on mobile */}
					{isMobile ? <BottomNav /> : null}
				</section>
			</div>
		);
	}
}

export default App;