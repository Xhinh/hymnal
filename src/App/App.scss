.App {
  width: 100%;
  text-align: center;

  &-body {
    background-color: #f5f5f5;
    color: white;
    display: flex;
    min-height: 100vh;
    flex-direction: column;
    font-size: calc(10px + 2vmin);
  }

  &-link {
    color: #61dafb;
  }
}