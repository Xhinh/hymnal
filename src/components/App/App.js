import React, { Component } from 'react';
import Nav from '../Nav/Nav';
import './App.scss';

class App extends Component {
  render() {
    return (
      <div className="App">
        <header className="App-body">
          <Nav />
        </header>
      </div>
    );
  }
}

export default App;
