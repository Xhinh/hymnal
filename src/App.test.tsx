import { Suspense } from "react";
import { BrowserRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import App from "./App";

jest.mock("__mocks__/createMocks.ts");
it("renders without crashing", () => {
	jest.spyOn(console, "log").mockImplementation(() => {});
	jest.spyOn(console, "warn").mockImplementation(() => {});
	window.scrollTo = jest.fn();
	const { asFragment } = render(
		<BrowserRouter>
			<Suspense fallback={<></>}>
				<App />
			</Suspense>
		</BrowserRouter>
	);
	expect(asFragment()).toMatchSnapshot();
});
