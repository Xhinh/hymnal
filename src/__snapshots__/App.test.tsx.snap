// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <section
      class="app_body"
    >
      <main
        class="app_inner "
      >
        <div
          class="wrapper"
        >
          <div
            class="css-1bsbzi3"
          >
            <div
              aria-valuemax="100"
              aria-valuemin="0"
              class="chakra-progress css-120wkjd"
              data-indeterminate=""
              role="progressbar"
            >
              <svg
                class="css-jf70f3"
                viewBox="0 0 100 100"
              >
                <circle
                  class="chakra-progress__track css-1bhxzgy"
                  cx="50"
                  cy="50"
                  r="42"
                  stroke-width="10px"
                />
                <circle
                  class="chakra-progress__indicator css-p6i8i8"
                  cx="50"
                  cy="50"
                  r="42"
                  stroke-width="10px"
                />
              </svg>
            </div>
          </div>
        </div>
      </main>
    </section>
  </div>
</DocumentFragment>
`;
